src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -540,10 +542,13 @@ internal void VerifyWrite(int originalDepth, Utf8JsonWriter writer)
         public abstract void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options);
 
         internal virtual T ReadWithQuotes(ref Utf8JsonReader reader)
-            => throw new InvalidOperationException();
+        {
+            ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
+            return default;
+        }
 
         internal virtual void WriteWithQuotes(Utf8JsonWriter writer, [DisallowNull] T value, JsonSerializerOptions options, ref WriteStack state)
-            => throw new InvalidOperationException();
+            => ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(TypeToConvert, this);
 
         internal sealed override void WriteWithQuotesAsObject(Utf8JsonWriter writer, object value, JsonSerializerOptions options, ref WriteStack state)
```

Comment:
Why do we box here?

Reply:
The only caller of this specific method is from `ObjectConverter`, when serializing `System.Object`-typed keys. The values to serialize are boxed at this point. `ObjectConverter` basically forwards the (boxed) value to the typed converter for the runtime type of the instance. The typed converter then unboxes the value and serializes it. 

It's the same as how `TryWriteAsObject` works.